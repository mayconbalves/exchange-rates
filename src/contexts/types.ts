export type ToastContextProps = {
  showToast: (message: string, type: 'success' | 'error') => void
}
