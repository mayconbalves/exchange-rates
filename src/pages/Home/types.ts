export type Registrations = {
  id: string
  admissionDate: string
  email: string
  employeeName: string
  status: 'REPROVED' | 'APPROVED' | 'REVIEW'
  cpf: string
}
