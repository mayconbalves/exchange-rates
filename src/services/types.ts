export type RateItem = {
  date: string
  rate: string
  value: number
  timestamp: number
}

export type ApiResponse = {
  success: boolean
  timestamp: number
  base: string
  date: string
  rates: Record<string, number>
}
