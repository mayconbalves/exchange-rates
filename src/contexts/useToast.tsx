import { createContext, ReactNode, useCallback, useContext, useState } from 'react'
import { ToastContextProps } from './types'

const ToastContext = createContext<ToastContextProps | undefined>(undefined)

export const useToast = () => {
  const context = useContext(ToastContext)
  if (!context) {
    throw new Error('useToast deve ser usado dentro de um ToastProvider')
  }
  return context
}

export const ToastProvider = ({ children }: { children: ReactNode }) => {
  const [toasts, setToasts] = useState<{ message: string; type: 'success' | 'error' }[]>([])

  const showToast = useCallback((message: string, type: 'success' | 'error') => {
    setToasts((prev) => [...prev, { message, type }])
    setTimeout(() => {
      setToasts((prev) => prev.slice(1))
    }, 3000)
  }, [])

  return (
    <ToastContext.Provider value={{ showToast }}>
      {children}
      <div style={{ position: 'fixed', bottom: '20px', right: '20px', zIndex: 999 }}>
        {toasts.map((toast, index) => (
          <div
            key={index}
            style={{
              marginBottom: '10px',
              padding: '10px 20px',
              borderRadius: '5px',
              color: '#fff',
              backgroundColor: toast.type === 'success' ? 'green' : 'red'
            }}
          >
            {toast.message}
          </div>
        ))}
      </div>
    </ToastContext.Provider>
  )
}
