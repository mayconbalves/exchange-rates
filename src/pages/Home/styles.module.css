.wrapper {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: calc(100vh - 144px);
  position: relative;
}

.wrapper__button {
  margin: 20px;
}
