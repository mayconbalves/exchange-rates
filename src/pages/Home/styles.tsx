import styled from 'styled-components'

export const Wrapper = styled.div`
  padding: 24px;
  display: flex;
  flex-direction: column;
`
