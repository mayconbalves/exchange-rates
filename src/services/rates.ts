import axios from 'axios'
import { logError } from '~/utils/logError'
import { ApiResponse, RateItem } from './types'

const API_URL = 'https://api.exchangeratesapi.io/v1/latest'

export const fetchRates = async () => {
  try {
    const response = await axios.get<ApiResponse>(API_URL, {
      params: {
        access_key: '02b8b5234e798123853a457bcfa10514'
      }
    })

    const data = response.data

    if (data.success) {
      const ratesArray: RateItem[] = Object.entries(data.rates).map(([rate, value]) => ({
        rate,
        value,
        date: data.date,
        timestamp: data.timestamp
      }))

      return ratesArray
    } else {
      logError('Erro ao carregar as taxas de câmbio: Resposta não bem-sucedida', data)
      return []
    }
  } catch (error) {
    logError('Erro ao carregar as taxas de câmbio:', error)
    return []
  }
}
