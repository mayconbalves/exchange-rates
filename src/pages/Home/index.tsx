import React, { useCallback, useEffect, useState } from 'react'
import { Button, Loader, Pagination, Table } from '~/components'
import { useToast } from '~/contexts/useToast'
import { fetchRates } from '~/services/rates'
import styles from './styles.module.css'
import { RateItem } from './types'

const Home: React.FC = () => {
  const [rates, setRates] = useState<RateItem[]>([])
  const [date, setDate] = useState<string | null>(null)
  const [load, setLoad] = useState<boolean>(false)
  const { showToast } = useToast()

  const [currentPage, setCurrentPage] = useState<number>(1)
  const itemsPerPage = 5

  const fetchAllRates = useCallback(async () => {
    setLoad(true)
    try {
      const data = await fetchRates()
      if (Array.isArray(data)) {
        setRates(data)
        setDate(data[0]?.date || null)
      }
      showToast('Sucesso ao trazer os dados.', 'success')
    } catch {
      showToast('Erro ao carregar registros.', 'error')
    } finally {
      setLoad(false)
    }
  }, [showToast])

  useEffect(() => {
    fetchAllRates()
  }, [fetchAllRates])

  const indexOfLastRate = currentPage * itemsPerPage
  const indexOfFirstRate = indexOfLastRate - itemsPerPage
  const currentRates = rates.slice(indexOfFirstRate, indexOfLastRate)

  return (
    <div className={styles.wrapper}>
      <h1>Os valores da tabela em relação ao Euro</h1>
      {rates.length > 0 && date && (
        <>
          <Table rates={currentRates} date={date} />

          <Pagination
            currentPage={currentPage}
            itemsPerPage={itemsPerPage}
            totalItems={rates.length}
            onPageChange={setCurrentPage}
          />
        </>
      )}
      <div className={styles.wrapper__button}>
        <Button onClick={fetchAllRates} title="Atualizar" />
      </div>
      {load && <Loader />}
    </div>
  )
}

export default Home
