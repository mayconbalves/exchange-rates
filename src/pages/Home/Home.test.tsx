import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { useToast } from '~/contexts/useToast'
import { fetchRates } from '~/services/rates'
import Home from '.'

jest.mock('~/services/rates')
jest.mock('~/contexts/useToast')

const mockRates = [
  { moeda: 'USD', valor: 1.12, horario: '2024-11-09T12:00:00Z' },
  { moeda: 'GBP', valor: 0.89, horario: '2024-11-09T12:00:00Z' }
]

describe('Home Component', () => {
  const mockFetchRates = fetchRates as jest.Mock
  const mockShowToast = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
    ;(useToast as jest.Mock).mockReturnValue({ showToast: mockShowToast })
  })

  it('should handle errors and show toast notification if data fetching fails', async () => {
    mockFetchRates.mockRejectedValue(new Error('Error fetching rates'))

    render(<Home />)

    await waitFor(() => {
      expect(mockShowToast).toHaveBeenCalledWith('Erro ao carregar registros.', 'error')
    })
  })

  it('should fetch new data when the refresh button is clicked', async () => {
    mockFetchRates.mockResolvedValue(mockRates)

    render(<Home />)

    await waitFor(() => expect(mockFetchRates).toHaveBeenCalledTimes(1))

    fireEvent.click(screen.getByRole('button', { name: /atualizar/i }))

    await waitFor(() => expect(mockFetchRates).toHaveBeenCalledTimes(2))
  })

  it('Should be create snapshot', () => {
    const { container } = render(<Home />)

    expect(container).toMatchSnapshot()
  })
})
