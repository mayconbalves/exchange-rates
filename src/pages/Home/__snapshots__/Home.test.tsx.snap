// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component Should be create snapshot 1`] = `
<div>
  <div
    class="wrapper"
  >
    <h1>
      Os valores da tabela em relação ao Euro
    </h1>
    <div
      class="wrapper__button"
    >
      <button
        class="button"
      >
        Atualizar
      </button>
    </div>
    <div
      class="wrapper"
    >
      <svg
        aria-busy="true"
        aria-label="rotating-lines-loading"
        class="sc-egkSDF bLjrOr"
        data-testid="rotating-lines-svg"
        role="progressbar"
        speed="0.75"
        stroke="#001b74"
        viewBox="0 0 48 48"
        width="96"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(0, 24, 24)"
          width="5"
        />
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(30, 24, 24)"
          width="5"
        />
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(60, 24, 24)"
          width="5"
        />
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(90, 24, 24)"
          width="5"
        />
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(120, 24, 24)"
          width="5"
        />
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(150, 24, 24)"
          width="5"
        />
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(180, 24, 24)"
          width="5"
        />
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(210, 24, 24)"
          width="5"
        />
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(240, 24, 24)"
          width="5"
        />
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(270, 24, 24)"
          width="5"
        />
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(300, 24, 24)"
          width="5"
        />
        <polyline
          class="sc-fAUdSK EuyAg"
          points="24,12 24,4"
          transform="rotate(330, 24, 24)"
          width="5"
        />
      </svg>
    </div>
  </div>
</div>
`;
